File: language/real/haskell/TransferV05.hs
module Main where

import Data.Char (toUpper)
import Data.List (isPrefixOf, nub, sort)
import qualified Data.Map.Strict as M
import qualified Data.Set as S
import System.Environment (getArgs)
import System.Exit (exitFailure)

data Status = Fail | Hold | Pass deriving (Eq, Ord, Show)
data Mode = FormalMediated | Nonformal deriving (Eq, Show)

data Packet = Packet
  { pid :: String
  , mode :: Maybe Mode
  , claimScope :: String
  , worldScope :: String
  , statementScope :: String
  , worldAuthority :: Maybe Status
  , scopeEdges :: [(String,String)]
  , witnesses :: [(Status,String)]
  , ancestors :: [(String,String,Status)]
  , preserved :: S.Set String
  , defeatStates :: [String]
  , formalCustody :: String
  , requested :: String
  } deriving Show

emptyPacket :: Packet
emptyPacket = Packet "" Nothing "" "" "" Nothing [] [] [] S.empty [] "" ""

upper :: String -> String
upper = map toUpper

parseStatus :: String -> Either String Status
parseStatus s = case upper s of
  "FAIL" -> Right Fail
  "HOLD" -> Right Hold
  "PASS" -> Right Pass
  x -> Left ("invalid status " ++ x)

statusText :: Status -> String
statusText Fail = "FAIL"
statusText Hold = "HOLD"
statusText Pass = "PASS"

stripQuotes :: String -> String
stripQuotes s
  | length s >= 2 && head s == '"' && last s == '"' = init (tail s)
  | otherwise = s

parsePacket :: String -> Either String Packet
parsePacket txt = go emptyPacket False False (zip [1..] (lines txt))
  where
    go p started ended []
      | not started = Left "missing REALTRANSFER 0.5"
      | not ended = Left "missing END"
      | otherwise = validateRequired p
    go p started ended ((n,raw):xs)
      | ended = go p started ended xs
      | null ws || "#" `isPrefixOf` head ws = go p started ended xs
      | head ws == "REALTRANSFER" =
          if ws == ["REALTRANSFER","0.5"]
          then go p True ended xs
          else Left ("line "++show n++": expected REALTRANSFER 0.5")
      | not started = Left ("line "++show n++": packet must begin REALTRANSFER 0.5")
      | head ws == "END" = go p started True xs
      | otherwise = step p ws >>= \p' -> go p' started ended xs
      where ws = words raw

    step p ws = case ws of
      ["id",x] -> Right p{pid=stripQuotes x}
      ["mode",x] -> case upper x of
        "FORMAL_MEDIATED" -> Right p{mode=Just FormalMediated}
        "NONFORMAL" -> Right p{mode=Just Nonformal}
        y -> Left ("invalid mode "++y)
      ["claim_scope",x] -> Right p{claimScope=x}
      ["world_scope",x] -> Right p{worldScope=x}
      ["statement_scope",x] -> Right p{statementScope=x}
      ["world_authority",x] -> do s<-parseStatus x; Right p{worldAuthority=Just s}
      ["scope_contains",a,b] -> Right p{scopeEdges=(a,b):scopeEdges p}
      ("semantic_witness":_:st:warrant:_) -> do
        s<-parseStatus st
        let w=upper warrant
        if w `elem` ["EXTERNAL","SELF","UNTESTED"]
          then Right p{witnesses=(s,w):witnesses p}
          else Left ("invalid warrant "++w)
      ["ancestor",a,role,st] -> do
        s<-parseStatus st
        let r=upper role
        if r `elem` ["LOAD_BEARING","AUXILIARY"]
          then Right p{ancestors=(a,r,s):ancestors p}
          else Left ("invalid ancestor role "++r)
      ["preserve",a] -> Right p{preserved=S.insert a (preserved p)}
      ("defeat_path":_:st:_) ->
        let d=upper st in
        if d `elem` ["REACHABLE","DEAD","FORBIDDEN"]
          then Right p{defeatStates=d:defeatStates p}
          else Left ("invalid defeat state "++d)
      ("formal_custody":st:_) ->
        let s=upper st in
        if s `elem` ["PASS","HOLD","FAIL","NOT_APPLICABLE"]
          then Right p{formalCustody=s}
          else Left ("invalid formal custody "++s)
      ["authorize",a] ->
        let x=upper a in
        if x `elem` ["PASS","HOLD","FAIL","NOT_APPLICABLE"]
          then Right p{requested=x}
          else Left ("invalid authorization "++x)
      _ -> Left ("malformed command: " ++ unwords ws)

    validateRequired p
      | null (pid p) || mode p == Nothing || null (claimScope p) || null (worldScope p)
        || worldAuthority p == Nothing || null (formalCustody p) || null (requested p)
        = Left "missing required transfer field"
      | mode p == Just FormalMediated && null (statementScope p)
        = Left "FORMAL_MEDIATED requires statement_scope"
      | otherwise = Right p

reachable :: [(String,String)] -> String -> String -> Bool
reachable edges sup sub
  | sup == sub = True
  | otherwise = bfs S.empty [sup]
  where
    bfs _ [] = False
    bfs seen (x:xs)
      | S.member x seen = bfs seen xs
      | otherwise =
          let ys=[b | (a,b)<-edges,a==x]
          in sub `elem` ys || bfs (S.insert x seen) (xs++ys)

rankText :: String -> Maybe Int
rankText "FAIL" = Just 0
rankText "HOLD" = Just 1
rankText "PASS" = Just 2
rankText _ = Nothing

derive :: Packet -> Either String (M.Map String String)
derive p = do
  wa <- maybe (Left "missing world authority") Right (worldAuthority p)
  let load=[(a,s)|(a,r,s)<-ancestors p,r=="LOAD_BEARING"]
  if null load then Left "no LOAD_BEARING ancestry" else Right ()
  let ceiling=minimum (map snd load)
  if wa > ceiling
    then Left ("WORLD_AUTHORITY_LAUNDERING: "++statusText wa++" exceeds load-bearing ancestry ceiling "++statusText ceiling)
    else Right ()
  case mode p of
    Just Nonformal -> do
      if formalCustody p /= "NOT_APPLICABLE" then Left "NONFORMAL requires formal_custody NOT_APPLICABLE" else Right ()
      if requested p /= "NOT_APPLICABLE" then Left "NONFORMAL transfer authorization must be NOT_APPLICABLE" else Right ()
      let sc=if reachable (scopeEdges p) (worldScope p) (claimScope p) then "PASS" else "FAIL"
      Right $ M.fromList
        [("semantic","NOT_APPLICABLE"),("scope",sc),("ancestry","NOT_APPLICABLE"),
         ("defeat","NOT_APPLICABLE"),("noncircular","NOT_APPLICABLE"),
         ("formal_custody","NOT_APPLICABLE"),("transfer","NOT_APPLICABLE")]
    Just FormalMediated -> do
      let sem=if not (null (witnesses p)) && all (\(s,w)->s==Pass && w=="EXTERNAL") (witnesses p) then "PASS" else "FAIL"
          sc=if reachable (scopeEdges p) (worldScope p) (claimScope p)
                && reachable (scopeEdges p) (statementScope p) (claimScope p) then "PASS" else "FAIL"
          anc=if all (\(a,_)->S.member a (preserved p)) load then "PASS" else "FAIL"
          def=if "REACHABLE" `elem` defeatStates p then "PASS" else "FAIL"
          nonc=if not (null (witnesses p)) && all ((/="SELF").snd) (witnesses p) then "PASS" else "FAIL"
          cust=case formalCustody p of
                 "PASS"->"PASS"; "HOLD"->"HOLD"; "FAIL"->"FAIL"; _->"FAIL"
          coords=[sem,sc,anc,def,nonc,cust]
          tr | "FAIL" `elem` coords || wa==Fail = "FAIL"
             | all (=="PASS") coords && wa==Pass = "PASS"
             | otherwise = "HOLD"
      rr <- maybe (Left "FORMAL_MEDIATED authorization may not be NOT_APPLICABLE") Right (rankText (requested p))
      gr <- maybe (Left "internal transfer status error") Right (rankText tr)
      if rr>gr then Left ("TRANSFER_LAUNDERING: requested "++requested p++" above derived "++tr) else Right ()
      Right $ M.fromList
        [("semantic",sem),("scope",sc),("ancestry",anc),("defeat",def),
         ("noncircular",nonc),("formal_custody",cust),("transfer",tr)]
    Nothing -> Left "missing mode"

emit :: Packet -> M.Map String String -> IO ()
emit p d = do
  putStrLn "HASKELL-REAL-TRANSFER=0.5"
  putStrLn ("id="++pid p)
  mapM_ (\k->putStrLn ("coordinate."++k++"="++M.findWithDefault "MISSING" k d))
    ["semantic","scope","ancestry","defeat","noncircular","formal_custody","transfer"]
  putStrLn "implementation_diversity_is_not_semantic_independence=true"

minimality :: IO ()
minimality = do
  let coords=["SEMANTIC","SCOPE","ANCESTRY","DEFEAT","NONCIRCULAR","FORMAL_CUSTODY"]
      attacks=M.fromList
        [("SEMANTIC_ALIAS","SEMANTIC"),("SCOPE_PROMOTION","SCOPE"),
         ("ANCESTRY_LOSS","ANCESTRY"),("DEFEAT_MIRAGE","DEFEAT"),
         ("BRIDGE_BOOTSTRAP","NONCIRCULAR"),("FORMAL_ROUTE_CUSTODY","FORMAL_CUSTODY")]
      subsets=filter (not.null) (power coords)
      blocks ss=all (`elem` ss) (M.elems attacks)
      good=filter blocks subsets
      m=minimum (map length good)
      mins=filter ((==m).length) good
  putStrLn ("FROZEN_ATTACK_COORDINATES="++show (length coords))
  putStrLn ("MINIMUM_BLOCKING_COORDINATES="++show m)
  putStrLn ("MINIMAL_SETS="++show (map sort mins))
  putStrLn "MINIMALITY_SCOPE=FROZEN_ATTACK_FAMILY_ONLY"
  where
    power []=[[]]
    power (x:xs)=let r=power xs in r++map (x:) r

main :: IO ()
main = do
  args<-getArgs
  case args of
    ["--minimality"] -> minimality
    [] -> putStrLn "usage: runghc TransferV05.hs PACKET | --minimality" >> exitFailure
    paths -> mapM_ run paths
  where
    run path = do
      txt<-readFile path
      case parsePacket txt >>= \p -> derive p >>= \d -> Right (p,d) of
        Left e -> putStrLn e >> exitFailure
        Right (p,d) -> emit p d

File: language/real/src/transfer_v05.rs
use std::{collections::{BTreeMap, BTreeSet}, env, fs, path::Path};

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
enum Status { Fail, Hold, Pass }

impl Status {
    fn parse(s: &str) -> Result<Self,String> {
        match s {
            "FAIL" => Ok(Self::Fail),
            "HOLD" => Ok(Self::Hold),
            "PASS" => Ok(Self::Pass),
            _ => Err(format!("invalid status {s}")),
        }
    }
    fn as_str(self) -> &'static str {
        match self { Self::Fail=>"FAIL", Self::Hold=>"HOLD", Self::Pass=>"PASS" }
    }
}

#[derive(Clone, Debug)]
struct SemanticWitness {
    id: String,
    status: Status,
    warrant: String,
    world_predicate: String,
    formal_predicate: String,
    criterion: String,
}

#[derive(Clone, Debug)]
struct Ancestor {
    id: String,
    role: String,
    status: Status,
}

#[derive(Clone, Debug)]
struct DefeatPath {
    id: String,
    state: String,
    receipt: String,
}

#[derive(Default)]
struct Packet {
    id: String,
    mode: String,
    claim_scope: String,
    world_scope: String,
    statement_scope: String,
    world_authority: Option<Status>,
    scope_edges: Vec<(String,String)>,
    witnesses: Vec<SemanticWitness>,
    ancestors: Vec<Ancestor>,
    preserved: BTreeSet<String>,
    defeat_paths: Vec<DefeatPath>,
    formal_custody: String,
    formal_receipt: String,
    requested: String,
}

fn tokens(s: &str) -> Result<Vec<String>,String> {
    let mut out=Vec::new();
    let mut cur=String::new();
    let mut q=false;
    let mut esc=false;
    for ch in s.chars() {
        if esc { cur.push(ch); esc=false; continue; }
        if ch=='\\' { esc=true; continue; }
        if ch=='"' { q=!q; continue; }
        if ch.is_whitespace() && !q {
            if !cur.is_empty() { out.push(std::mem::take(&mut cur)); }
        } else { cur.push(ch); }
    }
    if esc || q { return Err("unterminated quote/escape".into()); }
    if !cur.is_empty() { out.push(cur); }
    Ok(out)
}

fn reachable(edges:&[(String,String)], sup:&str, sub:&str) -> bool {
    if sup==sub { return true; }
    let mut seen=BTreeSet::new();
    let mut stack=vec![sup.to_string()];
    while let Some(x)=stack.pop() {
        if !seen.insert(x.clone()) { continue; }
        for (a,b) in edges {
            if a==&x {
                if b==sub { return true; }
                stack.push(b.clone());
            }
        }
    }
    false
}

fn parse(path:&Path)->Result<Packet,String>{
    let text=fs::read_to_string(path).map_err(|e|e.to_string())?;
    let mut p=Packet::default();
    let mut started=false;
    let mut ended=false;
    for (i,raw) in text.lines().enumerate() {
        let line=raw.trim();
        if line.is_empty() || line.starts_with('#') { continue; }
        let t=tokens(line).map_err(|e|format!("{}:{}: {e}",path.display(),i+1))?;
        if t.is_empty(){continue;}
        if t[0]=="REALTRANSFER" {
            if t.len()!=2 || t[1]!="0.5" { return Err("expected REALTRANSFER 0.5".into()); }
            started=true; continue;
        }
        if !started { return Err("packet must begin REALTRANSFER 0.5".into()); }
        if t[0]=="END" { ended=true; break; }
        match t[0].as_str() {
            "id" if t.len()==2 => p.id=t[1].clone(),
            "mode" if t.len()==2 => {
                let m=t[1].to_uppercase();
                if !["FORMAL_MEDIATED","NONFORMAL"].contains(&m.as_str()) { return Err(format!("invalid mode {m}")); }
                p.mode=m;
            }
            "claim_scope" if t.len()==2 => p.claim_scope=t[1].clone(),
            "world_scope" if t.len()==2 => p.world_scope=t[1].clone(),
            "statement_scope" if t.len()==2 => p.statement_scope=t[1].clone(),
            "world_authority" if t.len()==2 => p.world_authority=Some(Status::parse(&t[1].to_uppercase())?),
            "scope_contains" if t.len()==3 => p.scope_edges.push((t[1].clone(),t[2].clone())),
            "semantic_witness" if t.len()>=7 => {
                let status=Status::parse(&t[2].to_uppercase())?;
                let warrant=t[3].to_uppercase();
                if !["EXTERNAL","SELF","UNTESTED"].contains(&warrant.as_str()) { return Err(format!("invalid witness warrant {warrant}")); }
                p.witnesses.push(SemanticWitness{
                    id:t[1].clone(),status,warrant,
                    world_predicate:t[4].clone(),
                    formal_predicate:t[5].clone(),
                    criterion:t[6..].join(" "),
                });
            }
            "ancestor" if t.len()==4 => {
                let role=t[2].to_uppercase();
                if !["LOAD_BEARING","AUXILIARY"].contains(&role.as_str()) { return Err(format!("invalid ancestor role {role}")); }
                p.ancestors.push(Ancestor{id:t[1].clone(),role,status:Status::parse(&t[3].to_uppercase())?});
            }
            "preserve" if t.len()==2 => { p.preserved.insert(t[1].clone()); },
            "defeat_path" if t.len()>=4 => {
                let state=t[2].to_uppercase();
                if !["REACHABLE","DEAD","FORBIDDEN"].contains(&state.as_str()) { return Err(format!("invalid defeat state {state}")); }
                p.defeat_paths.push(DefeatPath{id:t[1].clone(),state,receipt:t[3..].join(" ")});
            }
            "formal_custody" if t.len()>=3 => {
                let state=t[1].to_uppercase();
                if !["PASS","HOLD","FAIL","NOT_APPLICABLE"].contains(&state.as_str()) { return Err(format!("invalid formal custody {state}")); }
                p.formal_custody=state;
                p.formal_receipt=t[2..].join(" ");
            }
            "authorize" if t.len()==2 => {
                let a=t[1].to_uppercase();
                if !["PASS","HOLD","FAIL","NOT_APPLICABLE"].contains(&a.as_str()) { return Err(format!("invalid authorization {a}")); }
                p.requested=a;
            }
            _ => return Err(format!("{}:{}: malformed command {line}",path.display(),i+1)),
        }
    }
    if !ended { return Err("missing END".into()); }
    if p.id.is_empty() || p.mode.is_empty() || p.claim_scope.is_empty() || p.world_scope.is_empty() || p.world_authority.is_none() || p.formal_custody.is_empty() || p.requested.is_empty() {
        return Err("missing required transfer field".into());
    }
    if p.mode=="FORMAL_MEDIATED" && p.statement_scope.is_empty() { return Err("FORMAL_MEDIATED requires statement_scope".into()); }
    Ok(p)
}

fn derived(p:&Packet)->Result<BTreeMap<&'static str,String>,String>{
    let mut out=BTreeMap::new();
    let world=p.world_authority.unwrap();

    let load:Vec<_>=p.ancestors.iter().filter(|a|a.role=="LOAD_BEARING").collect();
    if load.is_empty() { return Err("no LOAD_BEARING ancestry".into()); }
    let ancestry_ceiling=load.iter().map(|a|a.status).min().unwrap();
    if world>ancestry_ceiling {
        return Err(format!("WORLD_AUTHORITY_LAUNDERING: {} exceeds load-bearing ancestry ceiling {}",world.as_str(),ancestry_ceiling.as_str()));
    }

    if p.mode=="NONFORMAL" {
        if p.formal_custody!="NOT_APPLICABLE" { return Err("NONFORMAL requires formal_custody NOT_APPLICABLE".into()); }
        if p.requested!="NOT_APPLICABLE" { return Err("NONFORMAL transfer authorization must be NOT_APPLICABLE".into()); }
        out.insert("semantic","NOT_APPLICABLE".into());
        out.insert("scope", if reachable(&p.scope_edges,&p.world_scope,&p.claim_scope){"PASS"}else{"FAIL"}.into());
        out.insert("ancestry","NOT_APPLICABLE".into());
        out.insert("defeat","NOT_APPLICABLE".into());
        out.insert("noncircular","NOT_APPLICABLE".into());
        out.insert("formal_custody","NOT_APPLICABLE".into());
        out.insert("transfer","NOT_APPLICABLE".into());
        return Ok(out);
    }

    let semantic = if p.witnesses.is_empty() {
        "HOLD"
    } else if p.witnesses.iter().any(|w| w.status == Status::Fail) {
        "FAIL"
    } else if p.witnesses.iter().any(|w| w.status == Status::Hold) {
        "HOLD"
    } else {
        "PASS"
    };
    out.insert("semantic",semantic.into());

    let scope = if reachable(&p.scope_edges,&p.world_scope,&p.claim_scope)
        && reachable(&p.scope_edges,&p.statement_scope,&p.claim_scope) {"PASS"} else {"FAIL"};
    out.insert("scope",scope.into());

    let ancestry = if load.iter().all(|a|p.preserved.contains(&a.id)) {"PASS"} else {"FAIL"};
    out.insert("ancestry",ancestry.into());

    let defeat = if p.defeat_paths.iter().any(|d| d.state == "REACHABLE") {
        "PASS"
    } else if p.defeat_paths.iter().any(|d| d.state == "DEAD" || d.state == "FORBIDDEN") {
        "FAIL"
    } else {
        "HOLD"
    };
    out.insert("defeat",defeat.into());

    let noncircular = if p.witnesses.iter().any(|w| w.warrant == "SELF") {
        "FAIL"
    } else if p.witnesses.is_empty() || p.witnesses.iter().any(|w| w.warrant == "UNTESTED") {
        "HOLD"
    } else {
        "PASS"
    };
    out.insert("noncircular",noncircular.into());

    let custody = match p.formal_custody.as_str() {
        "PASS"=>"PASS",
        "HOLD"=>"HOLD",
        "FAIL"=>"FAIL",
        _=>"FAIL",
    };
    out.insert("formal_custody",custody.into());

    let coords=[semantic,scope,ancestry,defeat,noncircular,custody];
    let contract = if coords.iter().all(|x|*x=="PASS") && world==Status::Pass {
        "PASS"
    } else if coords.iter().any(|x|*x=="FAIL") || world==Status::Fail {
        "FAIL"
    } else {
        "HOLD"
    };
    out.insert("transfer",contract.into());

    let req_rank=match p.requested.as_str(){"FAIL"=>0,"HOLD"=>1,"PASS"=>2,_=>return Err("FORMAL_MEDIATED authorization may not be NOT_APPLICABLE".into())};
    let got_rank=match contract{"FAIL"=>0,"HOLD"=>1,"PASS"=>2,_=>0};
    if req_rank>got_rank { return Err(format!("TRANSFER_LAUNDERING: requested {} above derived {contract}",p.requested)); }

    Ok(out)
}

fn esc(s:&str)->String{
    s.replace('\\',"\\\\").replace('\n',"\\n").replace('=',"\\=")
}

fn canonical(p:&Packet,d:&BTreeMap<&'static str,String>)->String{
    let mut o=Vec::new();
    o.push("REAL-TRANSFER=0.5".into());
    o.push(format!("id={}",esc(&p.id)));
    o.push(format!("mode={}",p.mode));
    o.push(format!("world_authority={}",p.world_authority.unwrap().as_str()));
    o.push(format!("scope.claim={}",esc(&p.claim_scope)));
    o.push(format!("scope.world={}",esc(&p.world_scope)));
    o.push(format!("scope.statement={}",if p.statement_scope.is_empty(){"NOT_APPLICABLE".into()}else{esc(&p.statement_scope)}));
    for k in ["semantic","scope","ancestry","defeat","noncircular","formal_custody","transfer"] {
        o.push(format!("coordinate.{k}={}",d[k]));
    }
    o.push(format!("requested.authorization={}",p.requested));
    o.push(format!("witness.count={}",p.witnesses.len()));
    for (i,w) in p.witnesses.iter().enumerate() {
        o.push(format!("witness.{i}.id={}",esc(&w.id)));
        o.push(format!("witness.{i}.status={}",w.status.as_str()));
        o.push(format!("witness.{i}.warrant={}",w.warrant));
        o.push(format!("witness.{i}.world={}",esc(&w.world_predicate)));
        o.push(format!("witness.{i}.formal={}",esc(&w.formal_predicate)));
        o.push(format!("witness.{i}.criterion={}",esc(&w.criterion)));
    }
    o.push(format!("ancestor.count={}",p.ancestors.len()));
    for (i,a) in p.ancestors.iter().enumerate() {
        o.push(format!("ancestor.{i}.id={}",esc(&a.id)));
        o.push(format!("ancestor.{i}.role={}",a.role));
        o.push(format!("ancestor.{i}.status={}",a.status.as_str()));
        o.push(format!("ancestor.{i}.preserved={}",p.preserved.contains(&a.id)));
    }
    o.push(format!("defeat_path.count={}",p.defeat_paths.len()));
    for (i,x) in p.defeat_paths.iter().enumerate() {
        o.push(format!("defeat_path.{i}.id={}",esc(&x.id)));
        o.push(format!("defeat_path.{i}.state={}",x.state));
        o.push(format!("defeat_path.{i}.receipt={}",esc(&x.receipt)));
    }
    o.push(format!("formal_custody.receipt={}",esc(&p.formal_receipt)));
    o.push("transfer.meaning=STRUCTURAL_ADMISSIBILITY_NONAMPLIFYING".into());
    o.push("transfer.does_not_raise_world_authority=true".into());
    o.push("cross_language_agreement_is_not_world_correspondence=true".into());
    o.push("final_truth_distance=UNIDENTIFIED".into());
    o.join("\n")+"\n"
}

fn main(){
    let args:Vec<String>=env::args().skip(1).collect();
    if args.is_empty(){eprintln!("usage: real-v05-transfer PACKET...");std::process::exit(2);}
    let mut first=true;
    for a in args {
        match parse(Path::new(&a)).and_then(|p|derived(&p).map(|d|(p,d))) {
            Ok((p,d))=>{
                if !first{println!("---");}
                print!("{}",canonical(&p,&d));
                first=false;
            }
            Err(e)=>{eprintln!("{e}");std::process::exit(1);}
        }
    }
}
